shell/server/proxy.js

```diff
@@ -936,13 +937,69 @@ Proxy.prototype._callNewWebSession = function (request, userInfo) {
                                 WebSession.typeId, params).session;
 };
 
+var isRfc1918OrLocal = function(address) {
+  if (Net.isIPv4(address)) {
+    quad = address.split(".").map(function(x) { return parseInt(x, 10); });
+    return (quad[0] === 127 || quad[0] === 10 ||
+            (quad[0] === 192 && quad[1] === 168) ||
+            (quad[0] === 172 && quad[1] >= 16 && quad[1] < 32));
+  } else if (Net.isIPv6(address)) {
+    // IPv6 specifies ::1 as localhost and fd:: as reserved for private networks
+    return (address === "::1" || address.indexOf("fd") === 0);
+  } else {
+    // Ignore things that are neither IPv4 nor IPv6
+    return false;
+  }
+};
+
 Proxy.prototype._callNewApiSession = function (request, userInfo) {
   var self = this;
+  var params = {};
+
+  if ("x-sandstorm-passthrough" in request.headers) {
+    var optIns = request.headers["x-sandstorm-passthrough"]
+        .split(',')
+        .map(function(s) { return s.trim(); });
+    // The only currently supported passthrough value is "address", but others could be useful in
+    // the future
+    if (optIns.indexOf("address") !== -1) {
+      // Sadly, we can't use request.socket.remoteFamily because it's not available in the (rather-old)
+      // version of node that comes in the Meteor bundle we're using.  Hence this hackery.
+      var addressToPass = request.socket.remoteAddress;
+      if (isRfc1918OrLocal(addressToPass) && "x-real-ip" in request.headers) {
+        // Allow overriding the socket's remote address with X-Real-IP header if the request comes
+        // from either localhost or an RFC1918 address.  These are not useful for geolocation
+        // anyway.
+        addressToPass = request.headers["x-real-ip"];
+      }
+      if (Net.isIPv4(addressToPass)) {
+        // Map IPv4 addresses in IPv6
+        var v4Int = 0xFFFF00000000 + addressToPass.split(".")
```

Comment:
Maybe comment on how this comes out to 48 bits which manages to be precisely-representable in a double, hence you don't need to use bignum here?